interface_ospf: interface lookup performance fixes

chrisvanheuveln opened the pull request:
#### Summary
Part 2 of the interface lookup performance work.
Ref #630 

- Modified class method `InterfaceOspf.interfaces` to allow queries for single interfaces.
- `interface_ospf.yaml` is one of the early implementations and it was still using positional fields instead of named fields. This created a larger effort to implement the interface performance changes.
- `interface_ospf.rb` is the only consumer of the `InterfaceOspf` class.

chrisvanheuveln committed:
interface_ospf: interface lookup performance fixes

chrisvanheuveln committed:
fix assert msgs

chrisvanheuveln commented on lib/cisco_node_utils/interface_ospf.rb:
```diff
@@ -23,53 +23,62 @@
-      fail "interface #{int_name} does not exist" if @interface.nil?
+      @intf_name = intf_name.downcase
+      fail "interface #{@intf_name} does not exist" if
+        Interface.interfaces[@intf_name].nil?
```
This is probably a problem - because it's going to get all interfaces by default. Let me see if there's a solution here.

chrisvanheuveln replied:
Latest commit addresses this issue. You can continue the review now.

chrisvanheuveln committed:
api arg change to handle single_intf

mikewiebe commented on lib/cisco_node_utils/interface_ospf.rb:
```diff
@@ -143,109 +150,128 @@ def default_message_digest_encryption_type
+      set_args_keys_default
+    end
+
+    def destroy_hello_interval
```
Minor but can you combine `destroy_hello_interval` and `destroy_dead_interval` into one helper method instead of two?  Code is identical except for `dead_interval`|`hello_interval` string.

chrisvanheuveln replied:
done

chrisvanheuveln committed:
Merge branch 'develop' into intf-perf-fix

chrisvanheuveln committed:
destroy_* helpers combined to one method

chrisvanheuveln committed:
Add Interface.interface_count api

chrisvanheuveln committed:
Revert "Add Interface.interface_count api"

This reverts commit 62aa2b6c0a7f8367e5a870d989b47161adc943ba.

chrisvanheuveln committed:
Merge branch 'develop' into intf-perf-fix

chrisvanheuveln committed:
Interface.interfaces selection failure when 'interface' substrings present

Symptom:
- Interface output with an `'interface'` substring causes the `intf_list` to include invalid interface id's.
 - e.g. the following will cause `split` to create an erroneous `interface hsrp` from the line the follows the line with the substring on it
```
interface portchannel100
  hsrp bfd
  hsrp use-bia scope interface
                     ^^^^^^^^^
  hsrp mac-refresh 350
```

 - result:
```
pry(Cisco::Interface)> intf_list
=> [" port-channel100 hsrp bfd hsrp version 2 hsrp delay minimum 100 reload 200 hsrp use-bia scope ",
 " hsrp mac-refresh 350 ",
   ^^^^ (bug: this is now a separate entry in intf_list)
  ...
```

- Found with `cisco_hsrp/test_interface_hsrp.rb`

chrisvanheuveln commented:
I hosed the branch on this PR. Creating a new branch and abandoning this PR.
